### ISSUE TITLE
`lint-fmt` should use the same opam-repository hash as the builds

### ISSUE THREAD

**Issue by dra27** [ocurrent/ocaml-ci#350]
Both `lint-doc` and `lint-opam` reset opam-repository to the hash computed in the `analysis` stage. `lint-fmt` does not, which cause [failures in runs](https://ci.ocamllabs.io/github/ocaml/merlin/commit/7ba57970fff21a77ca3a18318bdcb78538e02825/variant/(lint-doc)) when a new Dune release is not yet available in the base image.

The `lint-fmt` stage (hopefully) simply needs to do the same:

```
RUN cd ~/opam-repository && (git cat-file -e 96c14a5143c10fe40487ddb2b446e5fe67d3b870 || git fetch origin master) && git reset -q --hard 96c14a5143c10fe40487ddb2b446e5fe67d3b870 && git log --no-decorate -n1 --oneline && opam update -u
```

step as the other stages.

**Comment by recoules** [comment]
I do not know if it is exactly related, but I got this error in the `lint-fmt` stage too when trying locally [ocaml-bitwuzla](https://github.com/bitwuzla/ocaml-bitwuzla):

```text
#13 [stage-0 5/5] RUN opam exec -- dune build @fmt || (echo "dune build @fmt failed"; exit 2)
#13 sha256:7828afe346e0766b85218e6b192bee48701a3bf7101186155367ef38fbee98ec
#13 0.599 File "dune", line 10, characters 8-14:
#13 0.599 10 |    (lib zarith)))
#13 0.599              ^^^^^^
#13 0.599 Error: Library "zarith" not found.
#13 0.599 Hint: try:
#13 0.599   dune external-lib-deps --missing @fmt
#13 0.600 dune build @fmt failed
#13 ERROR: executor failed running [/bin/sh -c opam exec -- dune build @fmt || (echo "dune build @fmt failed"; exit 2)]: exit code: 2
------
 > [stage-0 5/5] RUN opam exec -- dune build @fmt || (echo "dune build @fmt failed"; exit 2):
------
executor failed running [/bin/sh -c opam exec -- dune build @fmt || (echo "dune build @fmt failed"; exit 2)]: exit code: 2
2021-07-06 17:04.18: Job failed: Docker build exited with status 1
```
Here, the foreign stubs require the include directory of `zarith` but the `lint-fmt` did not install any of the dependencies, so the stage fails.

**Comment by talex5** [comment]
Fixed by @kit-ty-kate in #357.

@recoules that looks like a separate issue. Possibly a dune bug? I think `dune build @fmt` is supposed to work even if the dependencies aren't present.
